tlv/stream_test.go

```diff
@@ -0,0 +1,408 @@
+		name: "partial encode",
+		encGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{
+				u32: 3,
+				buf: bytes.Repeat([]byte{0x05}, 3),
+			}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(2, &o.u32),
+				tlv.MakePrimitiveRecord(4, &o.buf),
+			)
+			return o, s
+		},
+		decGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(2, &o.u32),
+				tlv.MakePrimitiveRecord(3, &o.u64),
+				tlv.MakePrimitiveRecord(4, &o.buf),
+			)
+			return o, s
+		},
+		expObj: &testObj{
+			u32: 3,
+			buf: bytes.Repeat([]byte{0x05}, 3),
+		},
+	},
+	{
+		name: "partial decode",
+		encGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{
+				u16: 2,
+				u32: 5,
+				u64: 7,
+				buf: bytes.Repeat([]byte{0x05}, 3),
+			}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(3, &o.u32),
+				tlv.MakePrimitiveRecord(5, &o.u64),
+				tlv.MakePrimitiveRecord(7, &o.buf),
+			)
+			return o, s
+		},
+		decGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(5, &o.u64),
+			)
+			return o, s
+		},
+		expObj: &testObj{
+			u16: 2,
+			u64: 7,
+		},
+	},
+	{
+		name: "full decode encode",
+		encGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{
+				u16: 2,
+				u32: 3,
+				u64: 7,
+				buf: bytes.Repeat([]byte{0x05}, 3),
+			}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(2, &o.u32),
+				tlv.MakePrimitiveRecord(3, &o.u64),
+				tlv.MakePrimitiveRecord(4, &o.buf),
+			)
+			return o, s
+		},
+		decGen: func() (*testObj, *tlv.Stream) {
+			o := &testObj{}
+			s := tlv.NewStream(
+				wtwire.WriteElement, wtwire.ReadElement,
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(2, &o.u32),
+				tlv.MakePrimitiveRecord(3, &o.u64),
+				tlv.MakePrimitiveRecord(4, &o.buf),
+			)
+			return o, s
+		},
+		expObj: &testObj{
+			u16: 2,
+			u32: 3,
+			u64: 7,
+			buf: bytes.Repeat([]byte{0x05}, 3),
+		},
+	},
+}
+
+func TestTLV(t *testing.T) {
+	for _, test := range tlvTests {
+		t.Run(test.name, func(t *testing.T) {
+			testTLV(t, test)
+		})
+	}
+}
+
+func testTLV(t *testing.T, test tlvTestCase) {
+	_, s1 := test.encGen()
+
+	var b bytes.Buffer
+	err := s1.Encode(&b)
+	if err != nil {
+		t.Fatalf("unable to encode tlv stream: %v", err)
+	}
+
+	obj2, s2 := test.decGen()
+	err = s2.Decode(bytes.NewReader(b.Bytes()))
+	if err != nil {
+		t.Fatalf("unable to decode tlv stream: %v", err)
+	}
+
+	if !reflect.DeepEqual(test.expObj, obj2) {
+		t.Fatalf("mismatch object, want: %v, got: %v",
+			test.expObj, obj2)
+	}
+}
+
+type thing struct {
```

Comment:
This thing thing, can it be moved to a separate benchmark file?

Reply:
removed entirely as it was just initial testing code